### PR TITLE
Add multi-objective Pareto front visualization for Optuna studies

### PR DESCRIPTION
**Pull request by tonypzy**
## Description

`py/visdom/integrations/optuna.py`
  - Add an `optuna-pareto-front` dashboard pane for completed Optuna studies with **two or three** objectives.
  - Use the configured objective names as Pareto-front axis labels.

`py/tests/unit/optuna_integration.py`
  - Add separate tests for two-objective and three-objective metric logging and Pareto-front generation.

`README.md`
  - Add a multi-objective Optuna usage example.
  - Document Pareto-front dashboard behavior.
  - Clarify that Optuna intermediate reporting and pruning apply only to single-objective studies.

## Motivation and Context

This PR adds changes to support pareto-front visualization for studies with two or three objectives.

## How Has This Been Tested?

Run the following code
```

import argparse
import math
import time
from collections import Counter

import optuna
import visdom

from visdom.integrations import OptunaCallback


OBJECTIVE_NAMES = ("accuracy", "latency_ms", "memory_mb")
OBJECTIVE_DIRECTIONS = ("maximize", "minimize", "minimize")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description=(
            "Optimize accuracy, latency, and memory and verify the Visdom "
            "three-objective Pareto-front pane."
        )
    )
    parser.add_argument("--server", default="http://localhost")
    parser.add_argument("--port", type=int, default=8097)
    parser.add_argument("--base-url", default="/")
    parser.add_argument(
        "--env",
        help="Dashboard environment; defaults to a unique timestamped name.",
    )
    parser.add_argument("--n-trials", type=int, default=30)
    parser.add_argument("--refresh-every", type=int, default=5)
    parser.add_argument("--seed", type=int, default=42)
    args = parser.parse_args()
    if args.n_trials < 3:
        parser.error("--n-trials must be at least 3")
    if args.refresh_every < 1:
        parser.error("--refresh-every must be at least 1")
    return args


def dashboard_url(viz: visdom.Visdom, env: str) -> str:
    root = "{}:{}{}".format(viz.server, viz.port, viz.base_url).rstrip("/")
    return "{}/env/{}".format(root, env)


def objective(trial: optuna.Trial) -> tuple[float, float, float]:
    """Return accuracy, latency, and memory with intentional trade-offs."""
    layers = trial.suggest_int("layers", 1, 6)
    width = trial.suggest_int("width", 32, 256, step=32)
    dropout = trial.suggest_float("dropout", 0.0, 0.5)
    learning_rate = trial.suggest_float("learning_rate", 1e-5, 1e-1, log=True)
    batch_size = trial.suggest_categorical("batch_size", [16, 32, 64, 128])

    capacity = layers * width
    accuracy = (
        0.70
        + 0.22 * (1.0 - math.exp(-capacity / 500.0))
        - 0.08 * abs(dropout - 0.20)
        - 0.02 * abs(math.log10(learning_rate) + 3.0)
    )
    latency_ms = 2.0 + capacity / 55.0 + batch_size / 128.0
    memory_mb = 96.0 + capacity * 0.60 + batch_size * 1.5
    return accuracy, latency_ms, memory_mb


def assert_study_results(study: optuna.Study, expected_trials: int) -> None:
    complete_trials = [
        trial for trial in study.trials if trial.state.name == "COMPLETE"
    ]
    if len(complete_trials) != expected_trials:
        raise AssertionError(
            "expected {} complete trials, got {}".format(
                expected_trials,
                len(complete_trials),
            )
        )
    if any(len(trial.values or ()) != 3 for trial in complete_trials):
        raise AssertionError(
            "every completed trial must contain three objective values"
        )
    if not study.best_trials:
        raise AssertionError("expected at least one trial on the Pareto front")


def main() -> None:
    args = parse_args()
    dashboard_env = args.env or "optuna_three_objectives_{}".format(int(time.time()))
    viz = visdom.Visdom(
        server=args.server,
        port=args.port,
        base_url=args.base_url,
        env=dashboard_env,
        use_incoming_socket=False,
        raise_exceptions=True,
    )
    if not viz.check_connection(timeout_seconds=3):
        raise SystemExit(
            "Cannot connect to Visdom at {}:{}. Start the server first with "
            "`python3 -m visdom.server`.".format(args.server, args.port)
        )

    callback = OptunaCallback(
        viz,
        dashboard_env=dashboard_env,
        objective_names=OBJECTIVE_NAMES,
        create_dashboard=True,
        refresh_every=args.refresh_every,
        raise_on_error=True,
    )
    study = optuna.create_study(
        study_name="accuracy-latency-memory",
        directions=OBJECTIVE_DIRECTIONS,
        sampler=optuna.samplers.NSGAIISampler(seed=args.seed),
    )
    study.optimize(objective, n_trials=args.n_trials, callbacks=[callback])

    if not callback.update_dashboard(study):
        raise RuntimeError("the final Optuna dashboard update failed")
    assert_study_results(study, args.n_trials)

    pareto_exists = viz.win_exists("optuna-pareto-front", env=dashboard_env)
    if str(pareto_exists).lower() != "true":
        raise AssertionError("the three-objective Pareto-front pane was not created")

    states = Counter(trial.state.name for trial in study.trials)
    print("\nThree-objective Optuna integration test passed.")
    print("Objectives: {}".format(", ".join(OBJECTIVE_NAMES)))
    print("Directions: {}".format(", ".join(OBJECTIVE_DIRECTIONS)))
    print("Trial states: {}".format(dict(states)))
    print("Pareto trials: {}".format(len(study.best_trials)))
    for trial in study.best_trials[:5]:
        print("  trial {}: {}".format(trial.number, tuple(trial.values)))
    print("Dashboard environment: {}".format(dashboard_env))
    print("Dashboard URL: {}".format(dashboard_url(viz, dashboard_env)))


if __name__ == "__main__":
    main()
```

## Screenshots (if appropriate):

https://github.com/user-attachments/assets/f0434447-7f8f-4e54-bb99-e46492890881

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code refactor or cleanup (changes to existing code for improved readability or performance)

## Checklist:

- [ ] I adapted the version number under `py/visdom/VERSION` according to [Semantic Versioning](https://semver.org/)
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Summary by Sourcery

Extend the Optuna integration with optional study dashboards and multi-objective Pareto-front visualization while preserving single-objective intermediate reporting and pruning support.

New Features:
- Add optional Optuna study dashboards with summary, trial, history, timeline, intermediate-value, parameter-importance, and two- or three-objective Pareto-front visualizations.
- Support objective-specific metric logging and Pareto-front axis labels for multi-objective studies.
- Refresh dashboards periodically during optimization and allow explicit final updates.

Bug Fixes:
- Preserve and display Optuna intermediate values for single-objective trials, including pruned trials.

Enhancements:
- Improve Optuna dashboard summaries with study statistics and links to relevant trial environments.
- Keep short Optuna trials visible in timeline visualizations without altering their recorded durations.
- Sanitize study and trial environment identifiers for safe Visdom URLs.

Build:
- Document Plotly as an optional dependency for Optuna dashboard visualizations.

Documentation:
- Add single- and multi-objective Optuna dashboard examples and document visualization, refresh, intermediate reporting, and pruning behavior.

Tests:
- Add unit coverage for intermediate-value logging, dashboard generation and error handling, and two- and three-objective Pareto-front visualization.

## Summary by CodeRabbit

- **New Features**
  - Added optional Optuna study dashboards with summary, hyperparameter, optimization history, timeline, intermediate-value, parameter-importance, and Pareto-front visualizations.
  - Added configurable automatic refresh intervals, manual updates, intermediate-value tracking, and per-trial dashboard links.
  - Added support for multi-objective studies and objective directions.

- **Bug Fixes**
  - Improved dashboard validation and error handling with warnings for unsupported or unavailable visualizations.

- **Documentation**
  - Expanded Optuna setup guidance, examples, visualization requirements, refresh behavior, and single- versus multi-objective limitations.

- **Tests**
  - Added coverage for dashboards, trial statuses, intermediate values, error handling, and multi-objective visualizations.

### CODE REVIEW

**Comment by tonypzy** [comment]
@vedansh-5 @Manik-Khajuria-5 @Jayantparashar10 @rajnisht7 Please review

File: py/visdom/integrations/optuna.py [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+                )
+                kwargs: dict[str, Any] = {"target_name": objective_name}
+                if target is not None:
+                    kwargs["target"] = target
+                history = plot_optimization_history(study, **kwargs)
+                history.update_layout(
+                    title="Optimization History — {}".format(objective_name)
+                )
+                suffix = "-{}".format(index) if multi_objective else ""
+                figures.append(("optuna-history{}".format(suffix), history))
```
[the remark this answers is not in the document]
Comment (tonypzy) [reply py/visdom/integrations/optuna.py:378]: Actually, optuna will generate an empty plot and show a warning. it will not raise an error or stop the optimization.

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy There are few issues with error handling and env ids, especially when the study name contains slash as a result optuna tests are currently not being picked up by CI. Can u please fix it

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy  Looks like the refresh every type check from #1757 was dropped here

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy Could we make sure failed dashboard updates still respect refresh instead of retrying on every trial?

**Review by Manik-Khajuria-5 — CHANGES_REQUESTED** [review]
@tonypzy Can u please resolve the conflicts

**Review by Jayantparashar10 — APPROVED** [review]

**Comment by Jayantparashar10** [comment]
@tonypzy  resolve conflicts

**Review by vedansh-5 — APPROVED** [review]

**Review by Manik-Khajuria-5 — APPROVED** [review]

**Review by rajnisht7 — APPROVED** [review]

**Review by Saksham-Sirohi — APPROVED** [review]
